pkg/util/tls/tls.go

```diff
@@ -13,42 +13,57 @@ import (
 
 // ClientConfig is the config for client TLS.
 type ClientConfig struct {
-	CertPath string `yaml:"tls_cert_path"`
-	KeyPath  string `yaml:"tls_key_path"`
-	CAPath   string `yaml:"tls_ca_path"`
+	CertPath           string `yaml:"tls_cert_path"`
+	KeyPath            string `yaml:"tls_key_path"`
+	CAPath             string `yaml:"tls_ca_path"`
+	ServerName         string `yaml:"tls_server_name"`
+	InsecureSkipVerify bool   `yaml:"tls_insecure_skip_verify"`
 }
 
 // RegisterFlagsWithPrefix registers flags with prefix.
 func (cfg *ClientConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.StringVar(&cfg.CertPath, prefix+".tls-cert-path", "", "TLS cert path for the client")
 	f.StringVar(&cfg.KeyPath, prefix+".tls-key-path", "", "TLS key path for the client")
 	f.StringVar(&cfg.CAPath, prefix+".tls-ca-path", "", "TLS CA path for the client")
+	f.StringVar(&cfg.ServerName, prefix+".tls-server-name", "", "Override the expected name on the server certificate")
+	f.BoolVar(&cfg.InsecureSkipVerify, prefix+".tls-insecure-skip-verify", false, "TLS skip certificate validation")
 }
 
 // GetTLSConfig initialises tls.Config from config options
 func (cfg *ClientConfig) GetTLSConfig() (*tls.Config, error) {
-	if cfg.CertPath != "" && cfg.KeyPath != "" && cfg.CAPath != "" {
-		clientCert, err := tls.LoadX509KeyPair(cfg.CertPath, cfg.KeyPath)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to load TLS certificate %s,%s", cfg.CertPath, cfg.KeyPath)
-		}
+	// no tls config given at all
+	if cfg.CertPath == "" && cfg.KeyPath == "" && cfg.CAPath == "" && !cfg.InsecureSkipVerify && cfg.ServerName == "" {
+		return nil, nil
+	}
+
+	config := &tls.Config{
+		InsecureSkipVerify: cfg.InsecureSkipVerify,
+		ServerName:         cfg.ServerName,
+	}
 
+	// read ca certificates
+	if cfg.CAPath != "" {
 		var caCertPool *x509.CertPool
 		caCert, err := ioutil.ReadFile(cfg.CAPath)
 		if err != nil {
 			return nil, errors.Wrapf(err, "error loading ca cert: %s", cfg.CAPath)
 		}
 		caCertPool = x509.NewCertPool()
 		caCertPool.AppendCertsFromPEM(caCert)
-		if len(clientCert.Certificate) > 0 && caCertPool != nil {
-			return &tls.Config{
-				InsecureSkipVerify: true,
-				Certificates:       []tls.Certificate{clientCert},
-				RootCAs:            caCertPool,
-			}, nil
+
+		config.RootCAs = caCertPool
+	}
+
+	// read client certificate
+	if cfg.CertPath != "" || cfg.KeyPath != "" {
```

Comment:
Is having no `cfg.CertPath` and `cfg.KeyPath` a valid case, or we should validate (above) that in order to enable TLS they're both required?

Reply:
I had the hope the LoadX509KeyPair would give a nice error, but that's not the case, I will add a better error there